data/lua/helper.lua

```diff
@@ -102,8 +102,8 @@ end
 -- Metatable that redirects access to wml.variable.proxy
 local proxy_var_mt = {
 	__metatable = "WML variables",
-	__index    = function(t, k) return wml.variable.proxy[k] end,
-	__newindex = function(t, k, v) wml.variable.proxy[k] = v end,
+	__index    = function(_, k) return wml.variable.proxy[k] end,
+	__newindex = function(_, k, v) wml.variable.proxy[k] = v end,
```

Comment:
If you want to rename all these `t` variables, in my opinion they should be called `self`, not `_`.

Reply:
I don't want to rename them really. I want to avoid dead code and unused variables. The table value is irrelevant for this metatable, and is never used. When you declare it as such, you both help automated code linting tools (like luacheck) and readability for yourself, once you get used to the convention.

Reply:
This is basically the same as here: https://github.com/wesnoth/wesnoth/pull/2380#discussion_r162595394

Reply:
More details:
http://lua-users.org/wiki/LuaStyleGuide
> The variable consisting of only an underscore "_" is commonly used as a placeholder when you want to ignore the variable: 
> `for _,v in ipairs(t) do print(v) end`

Reply:
That's fine in loop variables. I don't like it in function parameters, and I particularly don't like the `self` variable being called something other than `self` (I probably would rename these ones from `t` to `self` as well). If you can get luacheck to ignore unused variables called `_`, presumably you can also get it to ignore unused variables called `self`, right?

I just thought of another argument against the use of `_` though - Wesnoth has a separate convention on the meaning of that variable, as a marker for translatable strings. Perhaps we should do the opposite and ensure lua check _doesn't_ ignore cases of unused variables named `_`, because we have legitimate cases where they _are_ used.

Reply:
Regarding specific question:
> If you can get luacheck to ignore unused variables called _, presumably you can also get it to ignore unused variables called self, right?

I don't do that manually, `luacheck` ignores underscore variables by default. IDE does the same. I can't control it, it's Lua convention.

Reply:
Ah, indeed - it's bad if Lua code first assigns to `_`, then attempts to use a localizable string somewhere.

```
local _, useful_variable = some_function()
wesnoth.message(_("This is supposed to be a translatable string"))
```

Reply:
@jyrkive yes, use of underscores for ignoring variables and for translations are mutually exclusive. I've boldly chose translations to use a different name though.

Reply:
Uh, if you call your textdomains something other than _, the wmlxgettext tool won't see them. It's probably not a problem for the tool to ignore unused textdomain definitions, though. The problem is not ignoring unused `self`.

On that topic, I'm guessing you could just tweak this line to get it to work:

https://github.com/mpeterv/luacheck/blob/master/src/luacheck/check.lua#L97

(Assuming that's the same luacheck you're working with, though I'd be surprised if there's more than one by the same name.)

Reply:
@CelticMinstrel well, since `luacheck` is written in Lua, if you go for source code modification, then everything is possible of course. And patch further upstream releases if/when they come is possible too. And tell people to override their local installation with a bundled one.

I just wouldn't like to break conventions. I don't feel special enough with wesnoth to try do that. (I also don't see why `self` is any better than `t`, in fact I would prefer `_`, then `t`.)

So here's what I'm gonna do. I'll revert all unused variable changes, and will leave out the rest. Will also squash the commits and fix the title.

Reply:
EDIT: edited comment above ^

Reply:
You can leave in any unused variable changes that involve outright removing the variable though. Like if you had `local x = 5` and then x was never used, feel free to remove that line.